src/ring_theory/chain_of_divisors.lean

```diff
@@ -248,3 +317,113 @@ begin
+  {m p : associates M} {n : associates N} (hn : n ≠ 0)
+  (hp : p ∈ normalized_factors m) (d : set.Iic m ≃o set.Iic n) :
+  multiplicity p m = multiplicity ↑(d ⟨p, dvd_of_mem_normalized_factors hp⟩) n :=
+begin
+  refine le_antisymm (multiplicity_prime_le_multiplicity_image_by_factor_order_iso hp d) _,
+  suffices : multiplicity ↑(d ⟨p, dvd_of_mem_normalized_factors hp⟩) n ≤
+    multiplicity ↑(d.symm (d ⟨p, dvd_of_mem_normalized_factors hp⟩)) m,
+  { rw [d.symm_apply_apply ⟨p, dvd_of_mem_normalized_factors hp⟩, subtype.coe_mk] at this,
+    exact this },
+  letI := classical.dec_eq (associates N),
+  simpa only [subtype.coe_eta] using
+    (multiplicity_prime_le_multiplicity_image_by_factor_order_iso
+      (mem_normalized_factors_factor_order_iso_of_mem_normalized_factors hn hp d) d.symm),
+end
+
+end
+
+variables [unique (Mˣ)] [unique (Nˣ)]
+
+/-- The order isomorphism between the factors of `mk m` and the factors of `mk n` induced by a
+  bijection between the factors of `m` and the factors of `n` that preserves `∣`. -/
+@[simps]
+def mk_factor_order_iso_of_factor_dvd_equiv
+  {m : M} {n : N} (d : {l : M // l ∣ m} ≃ {l : N // l ∣ n}) (hd : ∀ l l',
+  ((d l) : N) ∣ (d l') ↔ (l : M) ∣ (l' : M)) :
+   set.Iic (associates.mk m) ≃o set.Iic (associates.mk n) :=
+{ to_fun := λ l, ⟨associates.mk (d ⟨associates_equiv_of_unique_units ↑l,
+    by { obtain ⟨x, hx⟩ := l, rw [subtype.coe_mk,  associates_equiv_of_unique_units_apply,
+      out_dvd_iff], exact hx } ⟩),
+    mk_le_mk_iff_dvd_iff.mpr (subtype.prop (d ⟨associates_equiv_of_unique_units ↑l, _ ⟩)) ⟩,
+  inv_fun := λ l, ⟨associates.mk (d.symm (⟨(associates_equiv_of_unique_units ↑l),
+      by obtain ⟨x, hx⟩ := l ; rw [subtype.coe_mk, associates_equiv_of_unique_units_apply,
+      out_dvd_iff] ; exact hx ⟩)),
+    mk_le_mk_iff_dvd_iff.mpr (subtype.prop (d.symm ⟨associates_equiv_of_unique_units ↑l, _ ⟩)) ⟩,
+  left_inv := λ ⟨l, hl⟩, by simp only [subtype.coe_eta, equiv.symm_apply_apply, subtype.coe_mk,
+    associates_equiv_of_unique_units_apply, mk_out, out_mk, normalize_eq],
+  right_inv := λ ⟨l, hl⟩, by simp only [subtype.coe_eta, equiv.apply_symm_apply, subtype.coe_mk,
+    associates_equiv_of_unique_units_apply, out_mk, normalize_eq, mk_out],
+  map_rel_iff' :=
+    by rintros ⟨a, ha⟩ ⟨b, hb⟩ ; simp only [equiv.coe_fn_mk, subtype.mk_le_mk,
+      associates.mk_le_mk_iff_dvd_iff, hd, subtype.coe_mk, associates_equiv_of_unique_units_apply,
+      out_dvd_iff, mk_out] }
+
+variables [unique_factorization_monoid M] [unique_factorization_monoid N]
+  [decidable_eq M]
+
+lemma mem_normalized_factors_factor_dvd_iso_of_mem_normalized_factors [decidable_eq N] {m p : M}
+  {n : N} (hm : m ≠ 0) (hn : n ≠ 0) (hp : p ∈ normalized_factors m)
+  (d : {l : M // l ∣ m} ≃ {l : N // l ∣ n})
+  (hd : ∀ l l', ((d l) : N) ∣ (d l') ↔ (l : M) ∣ (l' : M)) :
+    ↑(d ⟨p, dvd_of_mem_normalized_factors hp⟩) ∈ normalized_factors n :=
```

Comment:
can you maybe generalise these a little? I notice you had to make this originally for the associates version and now for the `|` version; if the answer is "not easily", don't worry about it